### PR TITLE
Make useEffect(async) warning more verbose

### PR DESCRIPTION
**Pull request by gaearon**
Adds a more detailed message. People still get tripped over this:

* https://github.com/facebook/react/issues/14326
* https://mobile.twitter.com/bardiarastin/status/1064144863388680192

New message:

screen shot 2018-11-26 at 3 53 12 pm

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberCommitWork.js [hunk 1]
```diff
@@ -336,9 +336,17 @@ function commitHookEffectList(
                 'useEffect function must return a cleanup function or ' +
                   'nothing.%s%s',
                 typeof destroy.then === 'function'
-                  ? ' Promises and useEffect(async () => ...) are not ' +
-                    'supported, but you can call an async function inside an ' +
-                    'effect.'
+                  ? '\n\nIt looks like you wrote useEffect(async () => ...) or returned a Promise. ' +
+                    'Instead, you may write an async function separately, ' +
```
Comment (sophiebits) [inline packages/react-reconciler/src/ReactFiberCommitWork.js]: no comma after separately please

File: packages/react-reconciler/src/ReactFiberCommitWork.js [hunk 2]
```diff
@@ -336,9 +336,17 @@ function commitHookEffectList(
...
+                    'and then call it from inside the effect:\n\n' +
+                    'async function fetchComment(commentId) {\n' +
+                    '  // You can await here\n' +
+                    '}\n\n' +
```
Comment (Kovensky) [inline packages/react-reconciler/src/ReactFiberCommitWork.js]: More of a question: would it be better to write this function inside the effect callback instead?

Supposedly, the engine would only have to evaluate the function declaration when the effect callback is called, but I'm not sure it'd help anyway as it'd still have to create the shared closure object.

At any rate, writing it outside the component entirely would definitely be better.

Comment (gaearon) [reply packages/react-reconciler/src/ReactFiberCommitWork.js]: I don't really have an opinion on this but it's nice to show it doesn't have to be inside.

**Comment by alippai** [comment]
Is there a tracking bug for data fetching solution (possibly with error handling / timeout included)? Just being curious whether it will fit other use-cases too e.g Fetch API streaming body or Web Worker postMessage

**Pull request merged**